File: MovieCatalog.cs
// MovieCatalog.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.Json;
using System.Text.Json.Serialization;

public record Movie(
    int Id,
    string Title,
    string Director,
    int Year,
    string Genre,
    bool Watched,
    int Rating,
    string AddedDate
);

public class MoviesData
{
    public List<Movie> Movies { get; set; } = new();
}

public class MovieCatalog
{
    private List<Movie> movies = new();
    private int nextId = 1;

    public IReadOnlyList<Movie> Movies => movies.AsReadOnly();

    public Movie AddMovie(string title, string director, int year, string genre, bool watched, int rating)
    {
        if (rating < 1 || rating > 10) throw new ArgumentException("Оценка должна быть от 1 до 10");
        if (year < 1900 || year > DateTime.Now.Year)
            throw new ArgumentException($"Год должен быть от 1900 до {DateTime.Now.Year}");
        if (string.IsNullOrWhiteSpace(title) || string.IsNullOrWhiteSpace(director))
            throw new ArgumentException("Название и режиссёр не могут быть пустыми");
        if (string.IsNullOrWhiteSpace(genre)) genre = "Неизвестно";
        var movie = new Movie(nextId, title, director, year, genre, watched, rating, DateTime.Now.ToString("yyyy-MM-dd"));
        movies.Add(movie);
        nextId++;
        return movie;
    }

    public Movie? FindMovie(int id) => movies.FirstOrDefault(m => m.Id == id);

    public bool EditMovie(int id, Dictionary<string, object> updates)
    {
        var old = FindMovie(id);
        if (old == null) return false;
        movies.Remove(old);
        string title = updates.ContainsKey("title") ? (string)updates["title"] : old.Title;
        string director = updates.ContainsKey("director") ? (string)updates["director"] : old.Director;
        int year = updates.ContainsKey("year") ? (int)updates["year"] : old.Year;
        string genre = updates.ContainsKey("genre") ? (string)updates["genre"] : old.Genre;
        bool watched = updates.ContainsKey("watched") ? (bool)updates["watched"] : old.Watched;
        int rating = updates.ContainsKey("rating") ? (int)updates["rating"] : old.Rating;
        var updated = new Movie(old.Id, title, director, year, genre, watched, rating, old.AddedDate);
        movies.Add(updated);
        return true;
    }

    public bool DeleteMovie(int id) => movies.RemoveAll(m => m.Id == id) > 0;

    public List<Movie> SearchMovies(string query)
    {
        var q = query.ToLower();
        return movies.Where(m => m.Title.ToLower().Contains(q) || m.Director.ToLower().Contains(q)).ToList();
    }

    public List<Movie> FilterByWatched(bool watched) => movies.Where(m => m.Watched == watched).ToList();

    public List<Movie> FilterByGenre(string genre) =>
        movies.Where(m => string.Equals(m.Genre, genre, StringComparison.OrdinalIgnoreCase)).ToList();

    public List<Movie> FilterByYear(int year) => movies.Where(m => m.Year == year).ToList();

    public Dictionary<string, object> GetStats()
    {
        int total = movies.Count;
        int watchedCount = FilterByWatched(true).Count;
        int unwatched = total - watchedCount;
        double avgRating = FilterByWatched(true).Any() ? FilterByWatched(true).Average(m => m.Rating) : 0;
        var genres = movies.GroupBy(m => m.Genre).ToDictionary(g => g.Key, g => g.Count());
        return new Dictionary<string, object>
        {
            ["total"] = total,
            ["watched"] = watchedCount,
            ["unwatched"] = unwatched,
            ["avg_rating"] = avgRating,
            ["genres"] = genres
        };
    }

    public void SaveToFile(string filename)
    {
        var data = new MoviesData { Movies = movies };
        var options = new JsonSerializerOptions { WriteIndented = true };
        string json = JsonSerializer.Serialize(data, options);
        File.WriteAllText(filename, json);
    }

    public void LoadFromFile(string filename)
    {
        if (!File.Exists(filename)) return;
        string json = File.ReadAllText(filename);
        var data = JsonSerializer.Deserialize<MoviesData>(json);
        if (data != null)
        {
            movies = data.Movies;
            nextId = movies.Any() ? movies.Max(m => m.Id) + 1 : 1;
        }
    }
}

public static class Program
{
    private static string ReadString(string prompt)
    {
        Console.Write(prompt);
        return Console.ReadLine()?.Trim() ?? "";
    }

    private static int ReadInt(string prompt)
    {
        while (true)
        {
            Console.Write(prompt);
            if (int.TryParse(Console.ReadLine(), out int result))
                return result;
            Console.WriteLine("Введите число.");
        }
    }

    private static bool ReadBool(string prompt)
    {
        while (true)
        {
            string input = ReadString(prompt);
            if (input == "1") return true;
            if (input == "0") return false;
            Console.WriteLine("Введите 1 или 0.");
        }
    }

    private static void PrintMovie(Movie movie)
    {
        string status = movie.Watched ? "✅ Просмотрен" : "⏳ Не просмотрен";
        Console.WriteLine($"#{movie.Id} - {movie.Title} ({movie.Year})");
        Console.WriteLine($"   Режиссёр: {movie.Director}, Жанр: {movie.Genre}");
        Console.WriteLine($"   {status}, Оценка: {movie.Rating}/10");
        Console.WriteLine($"   Добавлен: {movie.AddedDate}");
    }

    public static void Main()
    {
        var catalog = new MovieCatalog();
        try { catalog.LoadFromFile("movies_data.json"); }
        catch { Console.WriteLine("Не удалось загрузить данные."); }

        while (true)
        {
            Console.WriteLine("\n===== КАТАЛОГ ФИЛЬМОВ (C#) =====");
            Console.WriteLine("1. Добавить фильм");
            Console.WriteLine("2. Показать все фильмы");
            Console.WriteLine("3. Показать просмотренные фильмы");
            Console.WriteLine("4. Показать непросмотренные фильмы");
            Console.WriteLine("5. Найти фильмы по названию или режиссёру");
            Console.WriteLine("6. Отметить фильм как просмотренный");
            Console.WriteLine("7. Редактировать фильм");
            Console.WriteLine("8. Удалить фильм");
            Console.WriteLine("9. Показать статистику");
            Console.WriteLine("10. Сохранить в файл");
            Console.WriteLine("11. Загрузить из файла");
            Console.WriteLine("0. Выход");
            string choice = ReadString("Выберите действие: ");

            switch (choice)
            {
                case "0": return;
                case "1":
                    string title = ReadString("Название: ");
                    if (string.IsNullOrWhiteSpace(title)) { Console.WriteLine("Название не может быть пустым."); continue; }
                    string director = ReadString("Режиссёр: ");
                    if (string.IsNullOrWhiteSpace(director)) { Console.WriteLine("Режиссёр не может быть пустым."); continue; }
                    int year = ReadInt("Год выпуска: ");
                    string genre = ReadString("Жанр: ");
                    if (string.IsNullOrWhiteSpace(genre)) genre = "Неизвестно";
                    bool watched = ReadBool("Статус (1-просмотрен, 0-не просмотрен): ");
                    int rating = ReadInt("Оценка (1-10): ");
                    try
                    {
                        var movie = catalog.AddMovie(title, director, year, genre, watched, rating);
                        Console.WriteLine($"Фильм добавлен с ID {movie.Id}");
                    }
                    catch (Exception ex) { Console.WriteLine($"Ошибка: {ex.Message}"); }
                    break;
                case "2":
                    if (!catalog.Movies.Any()) Console.WriteLine("Нет фильмов.");
                    else foreach (var m in catalog.Movies) PrintMovie(m);
                    break;
                case "3":
                    var watchedMovies = catalog.FilterByWatched(true);
                    if (!watchedMovies.Any()) Console.WriteLine("Нет просмотренных фильмов.");
                    else foreach (var m in watchedMovies) PrintMovie(m);
                    break;
                case "4":
                    var unwatchedMovies = catalog.FilterByWatched(false);
                    if (!unwatchedMovies.Any()) Console.WriteLine("Нет непросмотренных фильмов.");
                    else foreach (var m in unwatchedMovies) PrintMovie(m);
                    break;
                case "5":
                    string query = ReadString("Введите часть названия или режиссёра: ");
                    var results = catalog.SearchMovies(query);
                    if (!results.Any()) Console.WriteLine("Фильмы не найдены.");
                    else foreach (var m in results) PrintMovie(m);
                    break;
                case "6":
                    int id = ReadInt("Введите ID фильма: ");
                    var movie = catalog.FindMovie(id);
                    if (movie == null) { Console.WriteLine("Фильм не найден."); continue; }
                    catalog.EditMovie(id, new Dictionary<string, object> { ["watched"] = true });
                    Console.WriteLine("Фильм отмечен как просмотренный.");
                    break;
                case "7":
                    int eid = ReadInt("Введите ID фильма для редактирования: ");
                    var old = catalog.FindMovie(eid);
                    if (old == null) { Console.WriteLine("Фильм не найден."); continue; }
                    Console.WriteLine("Оставьте поле пустым, чтобы не менять.");
                    string newTitle = ReadString($"Название ({old.Title}): ");
                    string newDirector = ReadString($"Режиссёр ({old.Director}): ");
                    string newYearStr = ReadString($"Год ({old.Year}): ");
                    string newGenre = ReadString($"Жанр ({old.Genre}): ");
                    string newWatchedStr = ReadString($"Статус (1-просмотрен, 0-не просмотрен) сейчас: {(old.Watched ? "1" : "0")}: ");
                    string newRatingStr = ReadString($"Оценка ({old.Rating}): ");
                    var updates = new Dictionary<string, object>();
                    if (!string.IsNullOrWhiteSpace(newTitle)) updates["title"] = newTitle;
                    if (!string.IsNullOrWhiteSpace(newDirector)) updates["director"] = newDirector;
                    if (!string.IsNullOrWhiteSpace(newYearStr))
                    {
                        if (int.TryParse(newYearStr, out int y)) updates["year"] = y;
                        else Console.WriteLine("Год должен быть числом, пропускаем.");
                    }
                    if (!string.IsNullOrWhiteSpace(newGenre)) updates["genre"] = newGenre;
                    if (!string.IsNullOrWhiteSpace(newWatchedStr)) updates["watched"] = newWatchedStr == "1";
                    if (!string.IsNullOrWhiteSpace(newRatingStr))
                    {
                        if (int.TryParse(newRatingStr, out int r)) updates["rating"] = r;
                        else Console.WriteLine("Оценка должна быть числом, пропускаем.");
                    }
                    if (catalog.EditMovie(eid, updates)) Console.WriteLine("Фильм обновлён.");
                    else Console.WriteLine("Ошибка обновления.");
                    break;
                case "8":
                    int delId = ReadInt("Введите ID фильма для удаления: ");
                    if (catalog.DeleteMovie(delId)) Console.WriteLine("Фильм удалён.");
                    else Console.WriteLine("Фильм не найден.");
                    break;
                case "9":
                    var stats = catalog.GetStats();
                    Console.WriteLine("\n=== СТАТИСТИКА ===");
                    Console.WriteLine($"Всего фильмов: {stats["total"]}");
                    Console.WriteLine($"Просмотрено: {stats["watched"]}");
                    Console.WriteLine($"Не просмотрено: {stats["unwatched"]}");
                    Console.WriteLine($"Средняя оценка (только просмотренные): {stats["avg_rating"]:F2}");
                    Console.WriteLine("По жанрам:");
                    var genres = (Dictionary<string, int>)stats["genres"];
                    foreach (var kv in genres) Console.WriteLine($"  {kv.Key}: {kv.Value}");
                    break;
                case "10":
                    try { catalog.SaveToFile("movies_data.json"); Console.WriteLine("Сохранено."); }
                    catch (Exception ex) { Console.WriteLine($"Ошибка: {ex.Message}"); }
                    break;
                case "11":
                    try { catalog.LoadFromFile("movies_data.json"); Console.WriteLine("Загружено."); }
                    catch (Exception ex) { Console.WriteLine($"Ошибка: {ex.Message}"); }
                    break;
                default: Console.WriteLine("Неизвестная команда."); break;
            }
        }
    }
}

File: MovieCatalog.java
// MovieCatalog.java
import java.io.*;
import java.nio.file.*;
import java.time.LocalDate;
import java.util.*;
import java.util.stream.Collectors;

record Movie(int id, String title, String director, int year, String genre, boolean watched, int rating, String addedDate) implements Serializable {}

class MoviesData implements Serializable {
    private static final long serialVersionUID = 1L;
    public List<Movie> movies;
}

class MovieCatalog implements Serializable {
    private static final long serialVersionUID = 1L;
    private List<Movie> movies = new ArrayList<>();
    private int nextId = 1;

    public Movie addMovie(String title, String director, int year, String genre, boolean watched, int rating) {
        if (rating < 1 || rating > 10) throw new IllegalArgumentException("Оценка должна быть от 1 до 10");
        if (year < 1900 || year > LocalDate.now().getYear())
            throw new IllegalArgumentException("Год должен быть от 1900 до " + LocalDate.now().getYear());
        if (title.isBlank() || director.isBlank())
            throw new IllegalArgumentException("Название и режиссёр не могут быть пустыми");
        if (genre.isBlank()) genre = "Неизвестно";
        Movie movie = new Movie(nextId, title, director, year, genre, watched, rating, LocalDate.now().toString());
        movies.add(movie);
        nextId++;
        return movie;
    }

    public Optional<Movie> findMovie(int id) {
        return movies.stream().filter(m -> m.id() == id).findFirst();
    }

    public boolean editMovie(int id, Map<String, Object> updates) {
        Optional<Movie> opt = findMovie(id);
        if (opt.isEmpty()) return false;
        Movie old = opt.get();
        movies.remove(old);
        String title = (String) updates.getOrDefault("title", old.title());
        String director = (String) updates.getOrDefault("director", old.director());
        int year = (int) updates.getOrDefault("year", old.year());
        String genre = (String) updates.getOrDefault("genre", old.genre());
        boolean watched = (boolean) updates.getOrDefault("watched", old.watched());
        int rating = (int) updates.getOrDefault("rating", old.rating());
        Movie updated = new Movie(old.id(), title, director, year, genre, watched, rating, old.addedDate());
        movies.add(updated);
        return true;
    }

    public boolean deleteMovie(int id) {
        return movies.removeIf(m -> m.id() == id);
    }

    public List<Movie> searchMovies(String query) {
        String q = query.toLowerCase();
        return movies.stream()
                .filter(m -> m.title().toLowerCase().contains(q) || m.director().toLowerCase().contains(q))
                .collect(Collectors.toList());
    }

    public List<Movie> filterByWatched(boolean watched) {
        return movies.stream().filter(m -> m.watched() == watched).collect(Collectors.toList());
    }

    public List<Movie> filterByGenre(String genre) {
        return movies.stream().filter(m -> m.genre().equalsIgnoreCase(genre)).collect(Collectors.toList());
    }

    public List<Movie> filterByYear(int year) {
        return movies.stream().filter(m -> m.year() == year).collect(Collectors.toList());
    }

    public Map<String, Object> getStats() {
        int total = movies.size();
        int watchedCount = filterByWatched(true).size();
        int unwatched = total - watchedCount;
        double avgRating = filterByWatched(true).stream().mapToInt(Movie::rating).average().orElse(0);
        Map<String, Integer> genres = new HashMap<>();
        movies.forEach(m -> genres.put(m.genre(), genres.getOrDefault(m.genre(), 0) + 1));
        Map<String, Object> stats = new HashMap<>();
        stats.put("total", total);
        stats.put("watched", watchedCount);
        stats.put("unwatched", unwatched);
        stats.put("avg_rating", avgRating);
        stats.put("genres", genres);
        return stats;
    }

    public void saveToFile(String filename) throws IOException {
        MoviesData data = new MoviesData();
        data.movies = new ArrayList<>(movies);
        try (ObjectOutputStream oos = new ObjectOutputStream(Files.newOutputStream(Paths.get(filename)))) {
            oos.writeObject(data);
        }
    }

    public void loadFromFile(String filename) throws IOException, ClassNotFoundException {
        try (ObjectInputStream ois = new ObjectInputStream(Files.newInputStream(Paths.get(filename)))) {
            MoviesData data = (MoviesData) ois.readObject();
            movies = new ArrayList<>(data.movies);
            for (Movie m : movies) {
                if (m.id() >= nextId) nextId = m.id() + 1;
            }
        }
    }

    public List<Movie> getMovies() { return Collections.unmodifiableList(movies); }
}

public class MovieCatalogApp {
    private static final Scanner scanner = new Scanner(System.in);

    private static String readString(String prompt) {
        System.out.print(prompt);
        return scanner.nextLine().trim();
    }

    private static int readInt(String prompt) {
        while (true) {
            try {
                System.out.print(prompt);
                return Integer.parseInt(scanner.nextLine().trim());
            } catch (NumberFormatException e) {
                System.out.println("Введите число.");
            }
        }
    }

    private static boolean readBool(String prompt) {
        while (true) {
            String input = readString(prompt);
            if (input.equals("1")) return true;
            if (input.equals("0")) return false;
            System.out.println("Введите 1 или 0.");
        }
    }

    private static void printMovie(Movie movie) {
        String status = movie.watched() ? "✅ Просмотрен" : "⏳ Не просмотрен";
        System.out.printf("#%d - %s (%d)%n", movie.id(), movie.title(), movie.year());
        System.out.printf("   Режиссёр: %s, Жанр: %s%n", movie.director(), movie.genre());
        System.out.printf("   %s, Оценка: %d/10%n", status, movie.rating());
        System.out.printf("   Добавлен: %s%n", movie.addedDate());
    }

    public static void main(String[] args) {
        MovieCatalog catalog = new MovieCatalog();
        try {
            catalog.loadFromFile("movies_data.ser");
        } catch (IOException | ClassNotFoundException e) {
            System.out.println("Не удалось загрузить данные.");
        }

        while (true) {
            System.out.println("\n===== КАТАЛОГ ФИЛЬМОВ (Java) =====");
            System.out.println("1. Добавить фильм");
            System.out.println("2. Показать все фильмы");
            System.out.println("3. Показать просмотренные фильмы");
            System.out.println("4. Показать непросмотренные фильмы");
            System.out.println("5. Найти фильмы по названию или режиссёру");
            System.out.println("6. Отметить фильм как просмотренный");
            System.out.println("7. Редактировать фильм");
            System.out.println("8. Удалить фильм");
            System.out.println("9. Показать статистику");
            System.out.println("10. Сохранить в файл");
            System.out.println("11. Загрузить из файла");
            System.out.println("0. Выход");
            String choice = readString("Выберите действие: ");

            switch (choice) {
                case "0" -> { return; }
                case "1" -> {
                    String title = readString("Название: ");
                    if (title.isBlank()) { System.out.println("Название не может быть пустым."); continue; }
                    String director = readString("Режиссёр: ");
                    if (director.isBlank()) { System.out.println("Режиссёр не может быть пустым."); continue; }
                    int year = readInt("Год выпуска: ");
                    String genre = readString("Жанр: ");
                    if (genre.isBlank()) genre = "Неизвестно";
                    boolean watched = readBool("Статус (1-просмотрен, 0-не просмотрен): ");
                    int rating = readInt("Оценка (1-10): ");
                    try {
                        Movie movie = catalog.addMovie(title, director, year, genre, watched, rating);
                        System.out.println("Фильм добавлен с ID " + movie.id());
                    } catch (Exception e) {
                        System.out.println("Ошибка: " + e.getMessage());
                    }
                }
                case "2" -> {
                    if (catalog.getMovies().isEmpty()) System.out.println("Нет фильмов.");
                    else catalog.getMovies().forEach(MovieCatalogApp::printMovie);
                }
                case "3" -> {
                    var movies = catalog.filterByWatched(true);
                    if (movies.isEmpty()) System.out.println("Нет просмотренных фильмов.");
                    else movies.forEach(MovieCatalogApp::printMovie);
                }
                case "4" -> {
                    var movies = catalog.filterByWatched(false);
                    if (movies.isEmpty()) System.out.println("Нет непросмотренных фильмов.");
                    else movies.forEach(MovieCatalogApp::printMovie);
                }
                case "5" -> {
                    String query = readString("Введите часть названия или режиссёра: ");
                    var results = catalog.searchMovies(query);
                    if (results.isEmpty()) System.out.println("Фильмы не найдены.");
                    else results.forEach(MovieCatalogApp::printMovie);
                }
                case "6" -> {
                    int id = readInt("Введите ID фильма: ");
                    var opt = catalog.findMovie(id);
                    if (opt.isEmpty()) { System.out.println("Фильм не найден."); continue; }
                    Movie old = opt.get();
                    catalog.editMovie(id, Map.of("watched", true));
                    System.out.println("Фильм отмечен как просмотренный.");
                }
                case "7" -> {
                    int id = readInt("Введите ID фильма для редактирования: ");
                    var opt = catalog.findMovie(id);
                    if (opt.isEmpty()) { System.out.println("Фильм не найден."); continue; }
                    Movie old = opt.get();
                    System.out.println("Оставьте поле пустым, чтобы не менять.");
                    String newTitle = readString("Название (" + old.title() + "): ");
                    String newDirector = readString("Режиссёр (" + old.director() + "): ");
                    String newYearStr = readString("Год (" + old.year() + "): ");
                    String newGenre = readString("Жанр (" + old.genre() + "): ");
                    String newWatchedStr = readString("Статус (1-просмотрен, 0-не просмотрен) сейчас: " + (old.watched() ? "1" : "0") + ": ");
                    String newRatingStr = readString("Оценка (" + old.rating() + "): ");
                    Map<String, Object> updates = new HashMap<>();
                    if (!newTitle.isBlank()) updates.put("title", newTitle);
                    if (!newDirector.isBlank()) updates.put("director", newDirector);
                    if (!newYearStr.isBlank()) {
                        try { updates.put("year", Integer.parseInt(newYearStr)); }
                        catch (NumberFormatException e) { System.out.println("Год должен быть числом, пропускаем."); }
                    }
                    if (!newGenre.isBlank()) updates.put("genre", newGenre);
                    if (!newWatchedStr.isBlank()) updates.put("watched", newWatchedStr.equals("1"));
                    if (!newRatingStr.isBlank()) {
                        try { updates.put("rating", Integer.parseInt(newRatingStr)); }
                        catch (NumberFormatException e) { System.out.println("Оценка должна быть числом, пропускаем."); }
                    }
                    if (catalog.editMovie(id, updates)) System.out.println("Фильм обновлён.");
                    else System.out.println("Ошибка обновления.");
                }
                case "8" -> {
                    int id = readInt("Введите ID фильма для удаления: ");
                    if (catalog.deleteMovie(id)) System.out.println("Фильм удалён.");
                    else System.out.println("Фильм не найден.");
                }
                case "9" -> {
                    var stats = catalog.getStats();
                    System.out.println("\n=== СТАТИСТИКА ===");
                    System.out.println("Всего фильмов: " + stats.get("total"));
                    System.out.println("Просмотрено: " + stats.get("watched"));
                    System.out.println("Не просмотрено: " + stats.get("unwatched"));
                    System.out.printf("Средняя оценка (только просмотренные): %.2f%n", stats.get("avg_rating"));
                    System.out.println("По жанрам:");
                    @SuppressWarnings("unchecked")
                    Map<String, Integer> genres = (Map<String, Integer>) stats.get("genres");
                    genres.forEach((g, c) -> System.out.println("  " + g + ": " + c));
                }
                case "10" -> {
                    try {
                        catalog.saveToFile("movies_data.ser");
                        System.out.println("Сохранено.");
                    } catch (IOException e) {
                        System.out.println("Ошибка сохранения: " + e.getMessage());
                    }
                }
                case "11" -> {
                    try {
                        catalog.loadFromFile("movies_data.ser");
                        System.out.println("Загружено.");
                    } catch (IOException | ClassNotFoundException e) {
                        System.out.println("Ошибка загрузки: " + e.getMessage());
                    }
                }
                default -> System.out.println("Неизвестная команда.");
            }
        }
    }
}
